sdk/core/azure-core/src/main/java/com/azure/core/models/CloudEvent.java

```diff
@@ -0,0 +1,470 @@
+        }
+        if (Objects.isNull(value)) {
+            throw LOGGER.logExceptionAsError(new IllegalArgumentException("'value' cannot be null."));
+        }
+        if (!validateAttributeName(name)) {
+            throw LOGGER.logExceptionAsError(new IllegalArgumentException(
+                "'name' must have only small-case alphanumeric characters and not be one of the CloudEvent reserved "
+                    + "attribute names"));
+        }
+        if (this.extensionAttributes == null) {
+            this.extensionAttributes = new HashMap<>();
+        }
+        this.extensionAttributes.put(name.toLowerCase(Locale.ENGLISH), value);
+        return this;
+    }
+
+    /**
+     * Get the spec version. Users don't need to access it because it's always 1.0.
+     * Make it package level to test deserialization.
+     * @return The spec version.
+     */
+    String getSpecVersion() {
+        return this.specVersion;
+    }
+
+    /**
+     * Set the spec version. Users don't need to access it because it's always 1.0.
+     * Make it package level to test serialization.
+     * @return the cloud event itself.
+     */
+    CloudEvent setSpecVersion(String specVersion) {
+        this.specVersion = specVersion;
+        return this;
+    }
+
+    private static boolean validateAttributeName(String name) {
+        if (RESERVED_ATTRIBUTE_NAMES.contains(name)) {
+            return false;
+        }
+        for (int i = 0; i < name.length(); i++) {
+            char c = name.charAt(i);
+            if (!((c >= 'a' && c <= 'z') || (c >= '0' && c <= '9'))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    static class JacksonSerializer implements JsonSerializer {
+        private final JacksonAdapter jacksonAdapter = new JacksonAdapter();
+
+        @Override
+        public <T> T deserialize(InputStream stream, TypeReference<T> typeReference) {
+            try {
+                return jacksonAdapter.deserialize(stream, typeReference.getJavaType(), SerializerEncoding.JSON);
+            } catch (IOException e) {
+                throw LOGGER.logExceptionAsError(new RuntimeException(e));
+            }
+        }
+
+        @Override
+        public <T> Mono<T> deserializeAsync(InputStream stream, TypeReference<T> typeReference) {
+            return Mono.defer(() -> Mono.just(deserialize(stream, typeReference)));
+        }
+
+        @Override
+        public void serialize(OutputStream stream, Object value) {
+            try {
+                jacksonAdapter.serialize(value, SerializerEncoding.JSON, stream);
+            } catch (IOException e) {
+                throw LOGGER.logExceptionAsError(new RuntimeException(e));
+            }
+        }
+
+        @Override
+        public Mono<Void> serializeAsync(OutputStream stream, Object value) {
+            return Mono.fromRunnable(() -> serialize(stream, value));
+        }
+
+        JacksonAdapter getJacksonAdapter() {
```

Comment:
Do we need this?

Reply:
Not really. This is all internal so I exposed it for my test code to use because I like its serialize method, which returns a String, better than JsonSerializer's serialize method, which needs to deal with OutputStream. I have no problem removing it since we move this inner class to a class in different package.